test/helpers/MarketHandler.sol

```diff
@@ -0,0 +1,293 @@
+/* solhint-disable contract-name-camelcase, func-name-mixedcase, not-rely-on-time */
+// SPDX-License-Identifier: MIT
+pragma solidity =0.8.17;
+import "@/test/helpers/market.sol";
+import "@/contracts/test/MockERC20Permit.sol";
+import {StdUtils} from "forge-std/StdUtils.sol";
+import {DecodedRemovalIdV0} from "@/contracts/RemovalIdLib.sol";
+import {AddressArrayLib, UInt256ArrayLib} from "@/contracts/ArrayLib.sol";
+
+using AddressArrayLib for address[];
+using UInt256ArrayLib for uint256[];
+
+contract MarketHandler is UpgradeableMarket, StdUtils {
+  using EnumerableSetUpgradeable for EnumerableSetUpgradeable.UintSet;
+
+  mapping(bytes32 => uint256) public numCalls;
+  // Handler "ghost variables" used to track state across randomized calls.
+  // These can be used while asserting invariants and also help govern the
+  // behavior of the random calls to create more meaningful cases.
+  EnumerableSetUpgradeable.UintSet allRemovalIds;
+  EnumerableSetUpgradeable.UintSet unlistedRemovalIds;
+  EnumerableSetUpgradeable.UintSet listedRemovalIds;
+  EnumerableSetUpgradeable.UintSet soldRemovalIds;
+  EnumerableSetUpgradeable.UintSet releasedIds;
+  mapping(uint256 => uint256) releasedAmounts;
+
+  // A bounded set of possible suppliers to use when minting removals
+  address[] allSuppliers = [
+    vm.addr(1),
+    vm.addr(2),
+    vm.addr(3),
+    vm.addr(4),
+    vm.addr(5),
+    vm.addr(6)
+  ];
+
+  bytes32 constant CREATE_CERTIFICATE_EVENT_SELECTOR =
+    keccak256(
+      "CreateCertificate(address,address,uint256,uint256,uint256[],uint256[],address,uint256,uint256)"
+    );
+  bytes32 constant UPDATE_CERTIFICATE_EVENT_SELECTOR =
+    keccak256(
+      "UpdateCertificate(uint256,uint256[],uint256[],uint256[],uint256[],address,uint256)"
+    );
+
+  /**
+   * @dev Mint a batch of removals incporporating randomness but minimizing reverts:
+   * - the batch size is between 1 and 10
+   * - the vintage is between 1990 and 2019
+   * - the subIdentifier is between 0 and 100 million
+   * - the supplier is chosen randomly from a predefined set of possible suppliers
+   * - the amounts are chosen randomly between 0 and 1 ether-
+   * - the projectId is between 1 and 1 billion
+   * - the scheduleStartTime is between 1 and intMax minus 10 years (avoid overflow when calculating end time for schedules)
+   * - the holdbackPercentage is between 0 and 100
+   * - other ID fields are hardcoded
+   */
+  function mintRemovalBatchMinimizeReverts(uint256 fuzz) external {
```

Comment:
I might be wrong here, but would it be easier to set up functions like this one by using bounded args instead of using a generic fuzz across a bunch of different variables?
So instead of one arg, (fuzz), you'd have (batchSize, supplierIndex, vintage, subIdentifier)... And then subIdentifier, for example, would look like `subIdentifier = bound(1, 100_000_000)`.

Reply:
so I thought carefully about this one and the reasons I ended up parameterizing these functions in the simplest possible way and then transforming the entropy internally were twofold:
1) most importantly, these functions may also be used as building blocks for other more complicated functions (such as how they're called in `endToEnd`).  when they're called indirectly like that, the fuzzer doesn't actually have access to their parameters, the calling function itself has to pass args.  so it's easier to pass around a single source of entropy than it is to have to parameterize the calling function with all the fuzzed params that might be required by its helpers.
2) some of these inputs require more complicated transformation than is offered by "bound" (such as modulation to achieve an index or something like that) so it seemed simpler to do all of those transformations internally.